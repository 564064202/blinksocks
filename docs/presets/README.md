# Presets

Presets are chaining and composable, built-in presets are listed here. If you want to customize a new preset, feel free to read [this](../development/guideline#preset) first.

## Built-In Presets

**functional**

* [stats](#stats)
* [tracker](#tracker)
* [access-control](#access-control)

**basic**

* [base-auth](#base-auth)*
* ~~[base-with-padding](#base-with-padding)*~~ (deprecated)
* ~~[base-auth-stream](#base-auth-stream)*~~ (deprecated)

**shadowsocks**

* [ss-base](#ss-base)*
* [ss-stream-cipher](#ss-stream-cipher)
* [ss-aead-cipher](#ss-aead-cipher)

**shadowsocksR**

* [ssr-auth-aes128-md5](#ssr-auth-aes128-md5)
* [ssr-auth-aes128-sha1](#ssr-auth-aes128-sha1)
* [ssr-auth-chain-a](#ssr-auth-chain-a)
* [ssr-auth-chain-b](#ssr-auth-chain-b)

**v2ray**

* [v2ray-vmess](#v2ray-vmess)*

**obfuscator**

* [obfs-random-padding](#obfs-random-padding)
* [obfs-http](#obfs-http)
* [obfs-tls1.2-ticket](#obfs-tls1.2-ticket)

**others**

* [aead-random-cipher](#aead-random-cipher)

> You **MUST** put preset signed with (*) to the presets list if you want to relay application data to multiple destinations.

## Use External Preset

If you installed blinksocks via **npm install -g blinksocks**, you are free to use external presets.

To use external presets, you can:

**Use public npm package:**

```
$ npm install -g blinksocks-preset-demo
```

```
"presets": [{"name": "blinksocks-preset-demo"}]
```

**Use private custom module:**

```
"presets": [{"name": "/path/to/your/preset.js"}]
```

> When use external preset, make sure that preset meets the requirements of the current Node.js environment.

To write your own preset, please refer to [Custom Preset](../development/custom-preset).

----

## [stats]

This preset perform statistics among traffic via this preset, you can put it anywhere in preset list to obtain **summary/instant/process** information of specific preset traffic. This preset has no side-effect to the traffic through it.

|     PARAMS      |         DESCRIPTION          | DEFAULT |
| :-------------- | :--------------------------- | :------ |
| save_to         | path to the result json file | -       |
| sample_interval | sample interval in seconds   | 30      |
| save_interval   | save interval in seconds     | 60      |

```
"presets": [{
  "name": "ss-base"
}, {
  "name": "ss-stream-cipher",
  "params": {
    "method": "aes-256-cfb"
  }
}, {
  "name": "stats",
  "params": {
    "save_to": "stats.json",
    "sample_interval": 1,
    "save_interval": 10
  }
}]
```

```
// stats.json
{
  "sample": {
    "from": 1503920985192,
    "to": 1503921026263,
    "duration": 41071
  },
  "summary": {
    "maxOutSpeed": 105254,
    "maxInSpeed": 158,
    "maxConnections": 2,
    "totalOutBytes": 130360,
    "totalOutPackets": 34,
    "totalInBytes": 314,
    "totalInPackets": 2,
    "totalBytes": 130674,
    "totalPackets": 36,
    "totalErrors": 0
  },
  "instant": {
    "outSpeed": 0,
    "inSpeed": 0,
    "totalConnections": 0,
    "errorRate": 0,
    "outBytesPerSecond": 3174.0157288597798,
    "outPacketsPerSecond": 0.8278347252319155,
    "inBytesPerSecond": 7.645297168318279,
    "inPacketsPerSecond": 0.04869616030775974,
    "totalBytesPerSecond": 3181.6610260280977,
    "totalPacketsPerSecond": 0.8765308855396753
  },
  "process": {
    "upTime": 42.067,
    "cpuUsage": {
      "user": 1219598,
      "system": 96237
    },
    "memoryUsage": {
      "rss": 54575104,
      "heapTotal": 28311552,
      "heapUsed": 21701240,
      "external": 3688374
    }
  }
}
```

## [tracker]

Track data send/receive events via this preset, and print a part of them after connection closed.

```
"presets": [
  ...
  {"name": "tracker"},
  ...
]
```

And you can get the track message in your terminal and log files:

```
[tracker] summary(out/in = 14/9, 5191b/3431b) abstract(127.0.0.1:55566 play.google.com:443 u 555 d 394 u 616 d 165 u 221 156 934 795 d 74 u 43 d 1201 u 51 174 531 d 51 573 u 51 172 841 d 51 854 u 51 d 68)
```

## [access-control]

Easy and powerful Access Control(ACL) to each connection.

> NOTE: This preset is for server use only.

|  PARAMS   |                          DESCRIPTION                          | DEFAULT |
| :-------- | :------------------------------------------------------------ | :------ |
| acl       | A path to a text file which contains a list of rules in order | -       |
| max_tries | The maximum tries from client                                 | 60      |

> NOTE: you'd better put this preset to the last of the preset list.

```
"presets": [
  ...,
  {
    "name": "access-control",
    "params": {
      "acl": "acl.txt",
      "max_tries": 60
    }
  }
]
```

**acl.txt** for example:

```
# [addr[/mask][:port]] [ban] [max_upload_speed(/s)] [max_download_speed(/s)]

example.com     1            # prevent access to example.com
example.com:*   1            # prevent access to example.com:*, equal to above
example.com:443 1            # prevent access to example.com:443 only
*:25            1            # prevent access to SMTP servers
*:*             1            # prevent all access from/to all endpoints
127.0.0.1       1            # ban localhost
192.168.0.0/16  1            # ban hosts in 192.168.*.*
172.27.1.100    0 120K       # limit upload speed to 120KB/s
172.27.1.100    0 -    120K  # limit download speed to 120KB/s
172.27.1.100    0 120K 120K  # limit upload and download speed to 120KB/s
```

Rules in **acl.txt** has a priority from lower to higher.

If server fail to process client(**by host**) requests over **max_tries** times, client will be **baned immediately**, and a new rule will be appended to the **acl** file.

To recovery unwary ban, you can edit acl file, remove unwanted rule without restarting the program.

> NOTE: rules will take effect immediately each time **acl.txt** was updated.

## [base-auth]

A preset based on "ss-base" and provides a HMAC to ensure integrity for addressing part.

| PARAMS |            DESCRIPTION             | DEFAULT |
| :----- | :--------------------------------- | :------ |
| method | a hash algorithm for creating HMAC | sha1    |

`method` can be one of:

md5, sha1, sha256

```
"presets": [{
  "name": "base-auth",
  "params": {
    "method": "sha1"
  }
}, {
  "name": "ss-stream-cipher",
  "params": {
    "method": "aes-256-cfb"
  }
}]
```

## ~~[base-with-padding]~~ (deprecated)

An advanced preset based on [ss-base], **SHOULD BE** used with ciphers in **cfb** operation mode.
It can prevent address from being tampered.

**NOTE**: Using [base-with-padding] with non-cfb ciphers will lose protection. 

| PARAMS |           DESCRIPTION           | DEFAULT |
| :----- | :------------------------------ | :------ |
| salt   | a string for generating padding | -       |

```
"presets": [{
  "name": "base-with-padding",
  "params": {
    "salt": "any string"
  }
}, {
  "name": "ss-stream-cipher",
  "params": {
    "method": "aes-256-cfb"
  }
}]
```

## ~~[base-auth-stream]~~ (deprecated)

A preset combines HMAC and stream encryption. HMAC only guarantees integrity for addressing part.

| PARAMS |           DESCRIPTION            | DEFAULT |
| :----- | :------------------------------- | :------ |
| method | encryption and decryption method | -       |

`method` can be one of:

aes-128-ctr, aes-192-ctr, aes-256-ctr,

aes-128-cfb, aes-192-cfb, aes-256-cfb,

camellia-128-cfb, camellia-192-cfb, camellia-256-cfb

```
"presets": [{
  "name": "base-auth-stream",
  "params": {
    "method": "aes-256-cfb"
  }
}]
```

## [ss-base]

This is a very basic preset which delivers the real destination address from client to server.

```
"presets": [
  {"name": "ss-base"}
]
```

## [ss-stream-cipher]

The shadowsocks's [stream cipher](https://shadowsocks.org/en/spec/Stream-Ciphers.html).

| PARAMS |           DESCRIPTION            | DEFAULT |
| :----- | :------------------------------- | :------ |
| method | encryption and decryption method | -       |

`method` can be one of:

aes-128-ctr, aes-192-ctr, aes-256-ctr,

aes-128-cfb, aes-192-cfb, aes-256-cfb,

camellia-128-cfb, camellia-192-cfb, camellia-256-cfb

```
"presets": [
  {
    "name": "ss-base"
  },
  {
    "name": "ss-stream-cipher",
    "params": {
      "method": "aes-256-ctr"
    }
  }
]
```

## [ss-aead-cipher]

The shadowsocks's [aead cipher](https://shadowsocks.org/en/spec/AEAD-Ciphers.html).

| PARAMS |           DESCRIPTION            | DEFAULT |
| :----- | :------------------------------- | :------ |
| method | encryption and decryption method | -       |

`method` can be one of:

aes-128-gcm, aes-192-gcm, aes-256-gcm,

chacha20-poly1305, chacha20-ietf-poly1305, xchacha20-ietf-poly1305

```
"presets": [
  {
    "name": "ss-base"
  },
  {
    "name": "ss-aead-cipher",
    "params": {
      "method": "aes-256-gcm"
    }
  }
]
```

## [ssr-auth-aes128-md5], [ssr-auth-aes128-sha1]

shadowsocksr [auth_aes128](https://github.com/shadowsocksr-rm/shadowsocks-rss/blob/master/doc/auth_aes128.md) implementation.

```
"presets": [
  {"name": "ss-base"},
  {"name": "ssr-auth-aes128-md5"},
  {"name": "ss-stream-cipher", "params": {"method": "aes-256-ctr"}}
]
```

## [ssr-auth-chain-a], [ssr-auth-chain-b]

shadowsocksr [auth_chain](https://github.com/shadowsocksr-rm/shadowsocks-rss/blob/master/doc/auth_chain_a.md) implementation.

```
"presets": [
  {"name": "ss-base"},
  {"name": "ssr-auth-chain-a"},
  {"name": "ss-stream-cipher", "params": {"method": "aes-256-ctr"}}
]
```

## [v2ray-vmess]

[v2ray vmess protocol](https://www.v2ray.com/chapter_04/03_vmess.html) implementation.

|  PARAMS  |            DESCRIPTION             |   DEFAULT   |
| :------- | :--------------------------------- | :---------- |
| id       | client uuid                        | -           |
| security | encryption method, **client only** | aes-128-gcm |

`method` can be one of:

aes-128-gcm, chacha20-poly1305, none

```
"presets": [
  {
    "name": "v2ray-vmess",
    "params": {
      "id": "c2485913-4e9e-41eb-8cc5-b2e7db8d3bc7",
      "security": "aes-128-gcm"
    }
  }
]
```

**Notice in v2ray configs:**

<details>
  <summary>v2ray client</summary>

  ```
    "outbound": {
      "protocol": "vmess",
      "settings": {
        "vnext": [
          {
            "address": "127.0.0.1",
            "port": 10086,
            "users": [
              {
                "id": "c2485913-4e9e-41eb-8cc5-b2e7db8d3bc7",
                "security": "aes-128-gcm",
                "alterId": 0 // [must be the default value: 0]
              }
            ]
          }
        ]
      },
      "mux": {
        "enabled": false // [must be false]
      }
    },
  ```

</details>

<details>
  <summary>v2ray server</summary>

```
  "inbound": {
    "port": 10086,
    "protocol": "vmess",
    "settings": {
      "clients": [
        {
          "id": "c2485913-4e9e-41eb-8cc5-b2e7db8d3bc7",
          "level": 1,
          "alterId": 0 // [must be the default value: 0]
        }
      ]
    }
  },
```

</details>

## [obfs-random-padding]

A simple obfuscator to significantly randomize the length of each packet. It can be used to prevent statistical analysis based on packet length.

```
"presets": [
  {
    "name": "ss-base"
  },
  {
    "name": "obfs-random-padding"
  },
  {
    "name": "ss-stream-cipher",
    "params": {
      "method": "aes-256-ctr"
    }
  }
]
```

## [obfs-http]

A http obfuscator, the first round after TCP handshake will wrap data within a random http header
selected from a text file.

| PARAMS |                  DESCRIPTION                  | DEFAULT |
| :----- | :-------------------------------------------- | :------ |
| file   | a text file which contains HTTP header paris. | -       |

`file` for example:

```
======================
GET / HTTP/1.1
Host: bing.com
Accept: */*
----------------------
HTTP/1.1 200 OK
Content-Type: text/plain
======================
POST /login HTTP/1.1
Host: login.live.com
Content-type: application/json
----------------------
HTTP/1.1 200 OK
======================
```

```
"presets": [
  {
    "name": "ss-base"
  },
  {
    "name": "obfs-http",
    "params": {
      "file": "path/to/fake.txt"
    }
  }
]
```

## [obfs-tls1.2-ticket]

A TLS1.2 obfuscator, do TLS handshake using SessionTicket TLS mechanism, transfer data inside of Application Data.

| PARAMS |                           DESCRIPTION                            | DEFAULT |
| :----- | :--------------------------------------------------------------- | :------ |
| sni    | [Server Name Indication], a server name or a list of server name | -       |

```
"presets": [
  {
    "name": "ss-base"
  },
  {
    "name": "obfs-tls1.2-ticket",
    "params": {
      "sni": ["cloudfront.net"]
    }
  }
]
```

## [aead-random-cipher]

This preset is based on **ss-aead-cipher**, but added random padding in the front of **each chunk**. This preset inherited
all features from **ss-aead-cipher** and prevent server from being detected by packet length statistics analysis.

|      PARAMS      |               DESCRIPTION                |   DEFAULT   |
| :--------------- | :--------------------------------------- | :---------- |
| method           | encryption and decryption method         | -           |
| info(optional)   | a string to generate subkey              | "bs-subkey" |
| factor(optional) | random padding length = (0-255) * factor | 2           |

```
"presets": [
  {
    "name": "base-with-padding",
    "params": {
      "salt": "any string"
    }
  },
  {
    "name": "aead-random-cipher",
    "params": {
      "method": "aes-256-gcm",
      "info": "bs-subkey",
      "factor": 2
    }
  }
]
```

## Have trouble in choosing presets?

Here is a [list](./RECOMMENDATIONS.md) of recommended conbinations.

[stats]: ../../src/presets/stats.js
[tracker]: ../../src/presets/tracker.js
[access-control]: ../../src/presets/access-control.js
[base-auth]: ../../src/presets/base-auth.js
[base-with-padding]: ../../src/presets/base-with-padding.js
[base-auth-stream]: ../../src/presets/base-auth-stream.js
[ss-base]: ../../src/presets/ss-base.js
[ss-stream-cipher]: ../../src/presets/ss-stream-cipher.js
[ss-aead-cipher]: ../../src/presets/ss-aead-cipher.js
[v2ray-vmess]: ../../src/presets/v2ray-vmess.js
[obfs-random-padding]: ../../src/presets/obfs-random-padding.js
[obfs-http]: ../../src/presets/obfs-http.js
[obfs-tls1.2-ticket]: ../../src/presets/obfs-tls1.2-ticket.js
[aead-random-cipher]: ../../src/presets/aead-random-cipher.js
[Server Name Indication]: https://en.wikipedia.org/wiki/Server_Name_Indication
