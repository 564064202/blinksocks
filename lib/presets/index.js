'use strict';Object.defineProperty(exports,'__esModule',{value:true});exports.presets=exports.getPresetClassByName=undefined;var _defs=require('./defs');Object.keys(_defs).forEach(function(key){if(key==='default'||key==='__esModule')return;Object.defineProperty(exports,key,{enumerable:true,get:function get(){return _defs[key]}})});var _proxy=require('./proxy');var _proxy2=_interopRequireDefault(_proxy);var _tunnel=require('./tunnel');var _tunnel2=_interopRequireDefault(_tunnel);var _stats=require('./stats');var _stats2=_interopRequireDefault(_stats);var _tracker=require('./tracker');var _tracker2=_interopRequireDefault(_tracker);var _accessControl=require('./access-control');var _accessControl2=_interopRequireDefault(_accessControl);var _expCompress=require('./exp-compress');var _expCompress2=_interopRequireDefault(_expCompress);var _expBaseWithPadding=require('./exp-base-with-padding');var _expBaseWithPadding2=_interopRequireDefault(_expBaseWithPadding);var _expBaseAuthStream=require('./exp-base-auth-stream');var _expBaseAuthStream2=_interopRequireDefault(_expBaseAuthStream);var _ssBase=require('./ss-base');var _ssBase2=_interopRequireDefault(_ssBase);var _ssStreamCipher=require('./ss-stream-cipher');var _ssStreamCipher2=_interopRequireDefault(_ssStreamCipher);var _ssAeadCipher=require('./ss-aead-cipher');var _ssAeadCipher2=_interopRequireDefault(_ssAeadCipher);var _aeadRandomCipher=require('./aead-random-cipher');var _aeadRandomCipher2=_interopRequireDefault(_aeadRandomCipher);var _obfsHttp=require('./obfs-http');var _obfsHttp2=_interopRequireDefault(_obfsHttp);var _obfsTls=require('./obfs-tls1.2-ticket');var _obfsTls2=_interopRequireDefault(_obfsTls);var _obfsRandomPadding=require('./obfs-random-padding');var _obfsRandomPadding2=_interopRequireDefault(_obfsRandomPadding);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj}}const mapping={'proxy':_proxy2.default,'tunnel':_tunnel2.default,'stats':_stats2.default,'tracker':_tracker2.default,'access-control':_accessControl2.default,'exp-compress':_expCompress2.default,'exp-base-with-padding':_expBaseWithPadding2.default,'exp-base-auth-stream':_expBaseAuthStream2.default,'ss-base':_ssBase2.default,'ss-stream-cipher':_ssStreamCipher2.default,'ss-aead-cipher':_ssAeadCipher2.default,'aead-random-cipher':_aeadRandomCipher2.default,'obfs-http':_obfsHttp2.default,'obfs-tls1.2-ticket':_obfsTls2.default,'obfs-random-padding':_obfsRandomPadding2.default};const presets=Object.keys(mapping);function getPresetClassByName(name){const clazz=mapping[name];if(clazz===undefined){throw Error(`cannot find preset: "${name}"`)}return clazz}exports.getPresetClassByName=getPresetClassByName;exports.presets=presets;