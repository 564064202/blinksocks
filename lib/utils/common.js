'use strict';Object.defineProperty(exports,'__esModule',{value:true});exports.BYTE_ORDER_LE=exports.BYTE_ORDER_BE=exports.ATYP_V6=exports.ATYP_DOMAIN=exports.ATYP_V4=undefined;exports.numberToBuffer=numberToBuffer;exports.parseURI=parseURI;exports.getRandomInt=getRandomInt;exports.getRandomChunks=getRandomChunks;exports.getChunks=getChunks;var _net=require('net');var _net2=_interopRequireDefault(_net);var _crypto=require('crypto');var _crypto2=_interopRequireDefault(_crypto);var _ip=require('ip');var _ip2=_interopRequireDefault(_ip);var _urijs=require('urijs');var _urijs2=_interopRequireDefault(_urijs);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj}}const ATYP_V4=exports.ATYP_V4=1;const ATYP_DOMAIN=exports.ATYP_DOMAIN=3;const ATYP_V6=exports.ATYP_V6=4;const BYTE_ORDER_BE=exports.BYTE_ORDER_BE=0;const BYTE_ORDER_LE=exports.BYTE_ORDER_LE=1;function numberToBuffer(num,len=2,byteOrder=BYTE_ORDER_BE){if(len<1){throw Error('len must be greater than 0')}const isOutOfRange=num>parseInt(`0x${'ff'.repeat(len)}`);if(isOutOfRange){throw Error(`Number ${num} is too big to put into a ${len} byte(s) size buffer`)}const buf=Buffer.alloc(len);if(byteOrder===BYTE_ORDER_BE){buf.writeUIntBE(num,0,len)}else{buf.writeUIntLE(num,0,len)}return buf}function parseURI(uri){let _uri=uri;let _port=null;if(_uri.startsWith('http://')){_uri=_uri.substr(7);_port=80}if(_uri.startsWith('https://')){_uri=_uri.substr(8);_port=443}const parts={};_urijs2.default.parseHost(_uri,parts);const hostname=parts.hostname,port=parts.port;const addrType=_net2.default.isIP(hostname)?_net2.default.isIPv4(hostname)?ATYP_V4:ATYP_V6:ATYP_DOMAIN;return{type:addrType,host:_net2.default.isIP(hostname)?_ip2.default.toBuffer(hostname):Buffer.from(hostname),port:numberToBuffer(port||_port||80)}}function getRandomInt(min,max){min=Math.ceil(min);max=Math.ceil(max);return Math.floor(_crypto2.default.randomBytes(1)[0]/255*(max-min))+min}function getRandomChunks(buffer,min,max){const totalLen=buffer.length;const bufs=[];let ptr=0;while(ptr<totalLen-1){const offset=getRandomInt(min,max);bufs.push(buffer.slice(ptr,ptr+offset));ptr+=offset}if(ptr<totalLen){bufs.push(buffer.slice(ptr))}return bufs}function getChunks(buffer,maxSize){const totalLen=buffer.length;const bufs=[];let ptr=0;while(ptr<totalLen-1){bufs.push(buffer.slice(ptr,ptr+maxSize));ptr+=maxSize}if(ptr<totalLen){bufs.push(buffer.slice(ptr))}return bufs}