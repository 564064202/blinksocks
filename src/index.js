import fs from 'fs';
import path from 'path';
import * as __modules__ from './core';

// only for webpack bundle
if (global.__WEBPACK__) {
  const argv = process.argv;
  const usage = 'Usage: node blinksocks.js -c/--config <json_file>';

  const conditions = [
    argv.length !== 4,
    (argv[2] !== '-c' && argv[2] !== '--config'),
    !argv[3].endsWith('.json')
  ];

  if (conditions.some((c) => c)) {
    console.log(usage);
    process.exit(0);
  }

  const file = path.resolve(process.cwd(), argv[3]);

  let json = null;
  try {
    const jsonFile = fs.readFileSync(file);
    json = JSON.parse(jsonFile);
  } catch (err) {
    throw Error(`fail to parse your '${file}': ${err.message}`);
  }

  const app = new __modules__.Hub(json);
  app.on('close', () => process.exit(0));
  app.run();

  process.on('SIGINT', () => app.terminate());
}

module.exports = __modules__;
