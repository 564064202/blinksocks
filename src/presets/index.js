import ProxyPreset from './proxy';
import TunnelPreset from './tunnel';
import StatsPreset from './stats';
import TrackerPreset from './tracker';
import ExpCompressPreset from './exp-compress';
import ExpBaseWithPaddingPreset from './exp-base-with-padding';
import ExpBaseAuthStreamPreset from './exp-base-auth-stream';
import SsBasePreset from './ss-base';
import SsStreamCipherPreset from './ss-stream-cipher';
import SsAeadCipherPreset from './ss-aead-cipher';
import AeadRandomCipherPreset from './aead-random-cipher';
import ObfsHttpPreset from './obfs-http';
import ObfsTls12TicketPreset from './obfs-tls1.2-ticket';
import ObfsRandomPaddingPreset from './obfs-random-padding';

const mapping = {
  'proxy': ProxyPreset,
  'tunnel': TunnelPreset,
  'stats': StatsPreset,
  'tracker': TrackerPreset,
  'exp-compress': ExpCompressPreset,
  'exp-base-with-padding': ExpBaseWithPaddingPreset,
  'exp-base-auth-stream': ExpBaseAuthStreamPreset,
  'ss-base': SsBasePreset,
  'ss-stream-cipher': SsStreamCipherPreset,
  'ss-aead-cipher': SsAeadCipherPreset,
  'aead-random-cipher': AeadRandomCipherPreset,
  'obfs-http': ObfsHttpPreset,
  'obfs-tls1.2-ticket': ObfsTls12TicketPreset,
  'obfs-random-padding': ObfsRandomPaddingPreset
};

const presets = Object.keys(mapping);

function getPresetClassByName(name) {
  const clazz = mapping[name];
  if (clazz === undefined) {
    throw Error(`cannot find preset: "${name}"`);
  }
  return clazz;
}

export {getPresetClassByName, presets};
export * from './defs';
