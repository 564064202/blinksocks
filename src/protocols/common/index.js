export * from './Message';
export * from './Constants';
