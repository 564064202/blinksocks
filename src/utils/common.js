import net from 'net';
import crypto from 'crypto';
import ip from 'ip';
import url from 'urijs';

export const ATYP_V4 = 1;
export const ATYP_DOMAIN = 3;
export const ATYP_V6 = 4;

export const BYTE_ORDER_BE = 0;
export const BYTE_ORDER_LE = 1;

/**
 * convert a number to a buffer with specified length in specified byte order
 * @param num
 * @param len
 * @param byteOrder
 * @returns {Buffer}
 */
export function numberToBuffer(num, len = 2, byteOrder = BYTE_ORDER_BE) {
  if (len < 1) {
    throw Error('len must be greater than 0');
  }

  const isOutOfRange = num > parseInt(`0x${'ff'.repeat(len)}`);
  if (isOutOfRange) {
    throw Error(`Number ${num} is too big to put into a ${len} byte(s) size buffer`);
  }

  const buf = Buffer.alloc(len);
  if (byteOrder === BYTE_ORDER_BE) {
    buf.writeUIntBE(num, 0, len);
  } else {
    buf.writeUIntLE(num, 0, len);
  }
  return buf;
}

/**
 * convert a http(s) url to an address with type, host and port
 * @param uri
 * @returns {{type: Number, host: Buffer, port: Buffer}}
 */
export function parseURI(uri) {
  let _uri = uri;
  let _port = null;

  if (_uri.startsWith('http://')) {
    _uri = _uri.substr(7);
    _port = 80;
  }

  if (_uri.startsWith('https://')) {
    _uri = _uri.substr(8);
    _port = 443;
  }

  const parts = {};
  url.parseHost(_uri, parts);

  const {hostname, port} = parts;
  const addrType = net.isIP(hostname) ? (net.isIPv4(hostname) ? ATYP_V4 : ATYP_V6) : ATYP_DOMAIN;

  return {
    type: addrType,
    host: net.isIP(hostname) ? ip.toBuffer(hostname) : Buffer.from(hostname),
    port: numberToBuffer(port || _port || 80)
  };
}

/**
 * returns a random integer in [min, max].
 * @param min
 * @param max
 * @returns {Number}
 */
export function getRandomInt(min, max) {
  min = Math.ceil(min);
  max = Math.ceil(max);
  return Math.floor(crypto.randomBytes(1)[0] / 0xff * (max - min)) + min;
}

/**
 * split buffer into chunks, each chunk size is picked randomly from [min, max]
 * @param buffer
 * @param min
 * @param max
 * @returns {Array<Buffer>}
 */
export function getRandomChunks(buffer, min, max) {
  const totalLen = buffer.length;
  const bufs = [];
  let ptr = 0;
  while (ptr < totalLen - 1) {
    const offset = getRandomInt(min, max);
    bufs.push(buffer.slice(ptr, ptr + offset));
    ptr += offset;
  }
  if (ptr < totalLen) {
    bufs.push(buffer.slice(ptr));
  }
  return bufs;
}

/**
 * split buffer into chunks, the max chunk size is maxSize
 * @param buffer
 * @param maxSize
 * @returns {Array<Buffer>}
 */
export function getChunks(buffer, maxSize) {
  const totalLen = buffer.length;
  const bufs = [];
  let ptr = 0;
  while (ptr < totalLen - 1) {
    bufs.push(buffer.slice(ptr, ptr + maxSize));
    ptr += maxSize;
  }
  if (ptr < totalLen) {
    bufs.push(buffer.slice(ptr));
  }
  return bufs;
}

/**
 * generate random id from [0, max - 1] but exclude ids
 * @param ids
 * @param max
 * @returns {Number}
 */
export function generateMutexId(ids = [], max = 0xff) {
  const seqArr = Array.from(new Array(max).keys());
  for (let i = 0; i < ids.length; ++i) {
    seqArr.splice(ids[i] - i, 1);
  }
  if (seqArr.length > 0) {
    return seqArr[crypto.randomBytes(1)[0] % seqArr.length];
  } else {
    return -1;
  }
}
