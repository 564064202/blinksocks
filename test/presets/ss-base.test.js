import {getPresetClassByName, CONNECT_TO_REMOTE} from '../../src/presets';
import {PresetRunner} from '../common';

test('running on client', async () => {
  const runner = new PresetRunner({
    clazz: getPresetClassByName('ss-base')
  }, {
    __IS_CLIENT__: true,
    __IS_SERVER__: false
  });

  runner.notify({
    type: CONNECT_TO_REMOTE,
    payload: {
      host: 'example.com',
      port: 443
    }
  });

  // should carry address info of above
  expect(await runner.forward('12')).toMatchSnapshot();
  // just payload
  expect(await runner.forward('34')).toMatchSnapshot();

  // just payload
  expect(await runner.backward('56')).toMatchSnapshot();

  runner.destroy();
});

test('running on server', async () => {
  const runner = new PresetRunner({
    clazz: getPresetClassByName('ss-base')
  }, {
    __IS_CLIENT__: false,
    __IS_SERVER__: true
  });

  const preset = runner.getPreset();
  preset.broadcast = jest.fn((action) => {
    expect(action).toMatchSnapshot();
    action.payload.onConnected();
  });

  // should decode address info correctly
  const header = [3, 11, 101, 120, 97, 109, 112, 108, 101, 46, 99, 111, 109, 1, 187];
  const payload = [0, 0, 0, 0];
  const ret = await runner.forward(Buffer.from(header.concat(payload)));
  expect(ret).toMatchSnapshot();
  expect(preset.broadcast).toHaveBeenCalled();
  // just payload
  expect(await runner.forward('12')).toMatchSnapshot();

  // just payload
  expect(await runner.backward('34')).toMatchSnapshot();
});
