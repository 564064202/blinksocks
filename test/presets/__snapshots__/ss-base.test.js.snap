// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`running on client 1`] = `
Object {
  "data": Array [
    3,
    11,
    101,
    120,
    97,
    109,
    112,
    108,
    101,
    46,
    99,
    111,
    109,
    1,
    187,
    49,
    50,
  ],
  "type": "Buffer",
}
`;

exports[`running on client 2`] = `
Object {
  "data": Array [
    51,
    52,
  ],
  "type": "Buffer",
}
`;

exports[`running on client 3`] = `
Object {
  "data": Array [
    53,
    54,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 1`] = `
Object {
  "payload": Object {
    "host": "0.0.0.0",
    "onConnected": [Function],
    "port": 443,
  },
  "type": "@action:connect_to_remote",
}
`;

exports[`running on server 2`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 3`] = `
Object {
  "data": Array [
    49,
    50,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 4`] = `
Object {
  "data": Array [
    51,
    52,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 5`] = `
Object {
  "payload": Object {
    "host": "::",
    "onConnected": [Function],
    "port": 443,
  },
  "type": "@action:connect_to_remote",
}
`;

exports[`running on server 6`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 7`] = `
Object {
  "data": Array [
    49,
    50,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 8`] = `
Object {
  "data": Array [
    51,
    52,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 9`] = `
Object {
  "payload": Object {
    "host": "example.com",
    "onConnected": [Function],
    "port": 443,
  },
  "type": "@action:connect_to_remote",
}
`;

exports[`running on server 10`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 11`] = `
Object {
  "data": Array [
    49,
    50,
  ],
  "type": "Buffer",
}
`;

exports[`running on server 12`] = `
Object {
  "data": Array [
    51,
    52,
  ],
  "type": "Buffer",
}
`;
